cirq/testing/circuit_compare.py

```diff
@@ -305,6 +308,75 @@ def assert_has_consistent_apply_unitary(
                                    atol=atol)
 
 
+def _assert_apply_unitary_works_when_axes_transposed(
+        val: Any,
+        *,
+        atol: float=1e-8) -> None:
+    """Tests whether a value's _apply_unitary_ handles out-of-order axes.
+
+    A common mistake to make when implementing `_apply_unitary_` is to assume
+    that the incoming axes will be contiguous, or ascending, or that they can be
+    flattened, or that other axes have a length of two, etc, etc ,etc. This
+    method checks that `_apply_unitary_` does the same thing to out-of-order
+    axes that it does to contiguous in-order axes.
+
+    Args:
+        val: The operation, gate, or other unitary object to test.
+        atol: Absolute error tolerance.
+    """
+
+    # Only test custom apply unitary methods.
+    if not hasattr(val, '_apply_unitary_') or not protocols.has_unitary(val):
+        return
+
+    # Pick sizes and shapes.
+    shape = protocols.qid_shape(val)
```

Comment:
This fails for schedules whose unitaries we appear to be testing, too. Either skip or make it work in this case(*). I think this is why the tests are failing.

(*) For example, by adding `_num_qubits_` or `_qid_shape_` to `Schedule`.

Reply:
I have expanded the PR to delete all the "qubit count" parameters out of the testing methods, and have them all use the qid_shape protocol instead. I fixed all test objects which did not specify this method. The tests predated the qid shape protocol existing, but now that it does we should of course use it.